packages/flutter/lib/src/material/paginated_data_table.dart

```diff
@@ -387,40 +387,43 @@ class PaginatedDataTableState extends State<PaginatedDataTable> {
       Container(width: 14.0),
     ]);
 
+    // The header will be visible only if the headerWidgets collection is not empty
+    final Semantics headerRow = Semantics(
+      container: true,
+      child: DefaultTextStyle(
+        // These typographic styles aren't quite the regular ones. We pick the closest ones from the regular
+        // list and then tweak them appropriately.
+        // See https://material.google.com/components/data-tables.html#data-tables-tables-within-cards
+        style: _selectedRowCount > 0 ? themeData.textTheme.subhead.copyWith(color: themeData.accentColor)
+            : themeData.textTheme.title.copyWith(fontWeight: FontWeight.w400),
+        child: IconTheme.merge(
+          data: const IconThemeData(
+              opacity: 0.54
+          ),
+          child: ButtonTheme.bar(
+            child: Ink(
+              height: 64.0,
+              color: _selectedRowCount > 0 ? themeData.secondaryHeaderColor : null,
+              child: Padding(
+                padding: EdgeInsetsDirectional.only(start: startPadding, end: 14.0),
+                child: Row(
+                    mainAxisAlignment: MainAxisAlignment.end,
+                    children: headerWidgets
+                ),
+              ),
+            ),
+          ),
+        ),
+      ),
+    );
+
     // CARD
     return Card(
       semanticContainer: false,
       child: Column(
         crossAxisAlignment: CrossAxisAlignment.stretch,
         children: <Widget>[
-          Semantics(
-            container: true,
-            child: DefaultTextStyle(
-              // These typographic styles aren't quite the regular ones. We pick the closest ones from the regular
-              // list and then tweak them appropriately.
-              // See https://material.io/design/components/data-tables.html#tables-within-cards
-              style: _selectedRowCount > 0 ? themeData.textTheme.subhead.copyWith(color: themeData.accentColor)
-                                           : themeData.textTheme.title.copyWith(fontWeight: FontWeight.w400),
-              child: IconTheme.merge(
-                data: const IconThemeData(
-                  opacity: 0.54
-                ),
-                child: ButtonTheme.bar(
-                  child: Ink(
-                    height: 64.0,
-                    color: _selectedRowCount > 0 ? themeData.secondaryHeaderColor : null,
-                    child: Padding(
-                      padding: EdgeInsetsDirectional.only(start: startPadding, end: 14.0),
-                      child: Row(
-                        mainAxisAlignment: MainAxisAlignment.end,
-                        children: headerWidgets
-                      ),
-                    ),
-                  ),
-                ),
-              ),
-            ),
-          ),
+          headerWidgets.isNotEmpty ? headerRow : Container(),
```

Comment:
rather than a Container(), if there is no header row, we should just not include the header.

e.g. create the List before the `return` statement, and build it with or without the header depending on whether you have one.

Reply:
Ok, the code isn't cleaner but we won't having ghost widgets in the tree so I'll rewrite it that way.